File: go/glyph/decimal128.go
package glyph

import (
	"errors"
	"fmt"
	"math/big"
	"regexp"
	"strconv"
	"strings"
)

// Decimal128 arithmetic errors
var (
	ErrDivisionByZero = errors.New("decimal128: division by zero")
	ErrOverflow       = errors.New("decimal128: overflow")
)

// Decimal128 represents a 128-bit decimal number: value = coefficient * 10^(-scale)
// where scale is -127 to 127 (8-bit signed) and coefficient is 16 bytes.
//
// This provides high-precision decimal arithmetic for financial, scientific,
// and blockchain applications where rounding errors are unacceptable.
type Decimal128 struct {
	Scale int8   // Exponent: -127 to 127
	Coef  [16]byte // 128-bit coefficient (two's complement, big-endian)
}

// NewDecimal128FromInt64 creates a Decimal128 from an int64.
func NewDecimal128FromInt64(value int64) Decimal128 {
	d := Decimal128{Scale: 0}
	coefInt := big.NewInt(value)
	coefBytes := coefInt.Bytes()
	
	// Handle negative numbers (two's complement)
	if value < 0 {
		// Convert to two's complement in 16 bytes
		coefInt.Add(coefInt, new(big.Int).Lsh(big.NewInt(1), 128))
		coefBytes = coefInt.Bytes()
	}
	
	// Copy to fixed 16-byte array
	if len(coefBytes) <= 16 {
		copy(d.Coef[16-len(coefBytes):], coefBytes)
	}
	
	return d
}

// NewDecimal128FromString creates a Decimal128 from a string representation.
// Examples: "123.45", "99.99", "0.0001234"
func NewDecimal128FromString(s string) (Decimal128, error) {
	s = strings.TrimSpace(s)
	
	// Find decimal point
	parts := strings.Split(s, ".")
	if len(parts) > 2 {
		return Decimal128{}, fmt.Errorf("invalid decimal format: %s", s)
	}
	
	d := Decimal128{}
	
	if len(parts) == 2 {
		// Has decimal part
		intPart := parts[0]
		fracPart := parts[1]
		
		// Scale is the number of fractional digits
		d.Scale = int8(len(fracPart))
		if d.Scale < -127 || d.Scale > 127 {
			return Decimal128{}, fmt.Errorf("scale out of range: %d", d.Scale)
		}
		
		// Combine into single integer
		coefStr := intPart + fracPart
		coefInt := new(big.Int)
		_, ok := coefInt.SetString(coefStr, 10)
		if !ok {
			return Decimal128{}, fmt.Errorf("invalid number: %s", s)
		}

		// Convert to 16 bytes
		coefBytes := intToCoef(coefInt)
		copy(d.Coef[:], coefBytes[:])
	} else {
		// No decimal part
		d.Scale = 0
		coefInt := new(big.Int)
		_, ok := coefInt.SetString(parts[0], 10)
		if !ok {
			return Decimal128{}, fmt.Errorf("invalid number: %s", s)
		}
		
		coefBytes := intToCoef(coefInt)
		copy(d.Coef[:], coefBytes[:])
	}

	return d, nil
}

// NewDecimal128FromFloat64 creates a Decimal128 from a float64.
// Note: Precision may be lost.
func NewDecimal128FromFloat64(f float64) (Decimal128, error) {
	return NewDecimal128FromString(strconv.FormatFloat(f, 'f', -1, 64))
}

// ToInt64 converts the decimal to int64 (truncates fractional part).
func (d Decimal128) ToInt64() int64 {
	coefInt := coefToInt(d.Coef[:])
	return coefInt.Int64()
}

// ToFloat64 converts the decimal to float64.
// Note: Precision may be lost.
func (d Decimal128) ToFloat64() float64 {
	coefInt := coefToInt(d.Coef[:])
	scale := new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(d.Scale)), nil)
	result := new(big.Float).SetInt(coefInt)
	scalef := new(big.Float).SetInt(scale)
	result.Quo(result, scalef)
	f, _ := result.Float64()
	return f
}

// String returns the string representation.
func (d Decimal128) String() string {
	coefInt := coefToInt(d.Coef[:])
	coefStr := coefInt.String()
	
	if d.Scale == 0 {
		return coefStr
	}
	
	// Handle negative
	negative := false
	if coefStr[0] == '-' {
		negative = true
		coefStr = coefStr[1:]
	}
	
	// Pad with zeros if needed
	for len(coefStr) < int(d.Scale)+1 {
		coefStr = "0" + coefStr
	}
	
	// Insert decimal point
	insertPos := len(coefStr) - int(d.Scale)
	result := coefStr[:insertPos] + "." + coefStr[insertPos:]
	
	if negative {
		result = "-" + result
	}
	
	return result
}

// Add returns d + other.
func (d Decimal128) Add(other Decimal128) (Decimal128, error) {
	coef1 := coefToInt(d.Coef[:])
	coef2 := coefToInt(other.Coef[:])

	targetScale := d.Scale
	if d.Scale != other.Scale {
		if d.Scale < other.Scale {
			scale := int64(other.Scale) - int64(d.Scale)
			coef1.Mul(coef1, new(big.Int).Exp(big.NewInt(10), big.NewInt(scale), nil))
			targetScale = other.Scale
		} else {
			scale := int64(d.Scale) - int64(other.Scale)
			coef2.Mul(coef2, new(big.Int).Exp(big.NewInt(10), big.NewInt(scale), nil))
		}
	}

	coef1.Add(coef1, coef2)

	if coef1.BitLen() > 127 {
		return Decimal128{}, ErrOverflow
	}

	result := Decimal128{Scale: targetScale}
	coefResult := intToCoef(coef1)
	copy(result.Coef[:], coefResult[:])
	return result, nil
}

// Sub returns d - other.
func (d Decimal128) Sub(other Decimal128) (Decimal128, error) {
	coef2 := coefToInt(other.Coef[:])
	coef2.Neg(coef2)
	negOther := Decimal128{Scale: other.Scale}
	negCoef := intToCoef(coef2)
	copy(negOther.Coef[:], negCoef[:])
	return d.Add(negOther)
}

// Mul returns d * other.
func (d Decimal128) Mul(other Decimal128) (Decimal128, error) {
	coef1 := coefToInt(d.Coef[:])
	coef2 := coefToInt(other.Coef[:])
	coef1.Mul(coef1, coef2)

	if coef1.BitLen() > 127 {
		return Decimal128{}, ErrOverflow
	}

	result := Decimal128{Scale: d.Scale + other.Scale}
	mulCoef := intToCoef(coef1)
	copy(result.Coef[:], mulCoef[:])
	return result, nil
}

// Div returns d / other.
func (d Decimal128) Div(other Decimal128) (Decimal128, error) {
	coef1 := coefToInt(d.Coef[:])
	coef2 := coefToInt(other.Coef[:])

	if coef2.Sign() == 0 {
		return Decimal128{}, ErrDivisionByZero
	}

	coef1.Div(coef1, coef2)

	result := Decimal128{Scale: d.Scale - other.Scale}
	divCoef := intToCoef(coef1)
	copy(result.Coef[:], divCoef[:])
	return result, nil
}

// Abs returns the absolute value.
func (d Decimal128) Abs() Decimal128 {
	coefInt := coefToInt(d.Coef[:])
	if coefInt.Sign() < 0 {
		coefInt.Neg(coefInt)
		coefBytes := intToCoef(coefInt)
		copy(d.Coef[:], coefBytes[:])
	}
	return d
}

// Neg returns the negation.
func (d Decimal128) Neg() Decimal128 {
	coefInt := coefToInt(d.Coef[:])
	coefInt.Neg(coefInt)
	coefBytes := intToCoef(coefInt)
	copy(d.Coef[:], coefBytes[:])
	return d
}

// Cmp compares two decimals. Returns -1 if d < other, 0 if d == other, 1 if d > other.
func (d Decimal128) Cmp(other Decimal128) int {
	d1 := coefToInt(d.Coef[:])
	d2 := coefToInt(other.Coef[:])
	
	// Align scales
	if d.Scale != other.Scale {
		if d.Scale < other.Scale {
			scale := int64(other.Scale) - int64(d.Scale)
			d1.Mul(d1, new(big.Int).Exp(big.NewInt(10), big.NewInt(scale), nil))
		} else {
			scale := int64(d.Scale) - int64(other.Scale)
			d2.Mul(d2, new(big.Int).Exp(big.NewInt(10), big.NewInt(scale), nil))
		}
	}
	
	return d1.Cmp(d2)
}

// Equal returns true if d == other.
func (d Decimal128) Equal(other Decimal128) bool {
	return d.Cmp(other) == 0
}

// LessThan returns true if d < other.
func (d Decimal128) LessThan(other Decimal128) bool {
	return d.Cmp(other) < 0
}

// GreaterThan returns true if d > other.
func (d Decimal128) GreaterThan(other Decimal128) bool {
	return d.Cmp(other) > 0
}

// IsZero returns true if d == 0.
func (d Decimal128) IsZero() bool {
	return d.Cmp(NewDecimal128FromInt64(0)) == 0
}

// IsNegative returns true if d < 0.
func (d Decimal128) IsNegative() bool {
	coefInt := coefToInt(d.Coef[:])
	return coefInt.Sign() < 0
}

// IsPositive returns true if d > 0.
func (d Decimal128) IsPositive() bool {
	coefInt := coefToInt(d.Coef[:])
	return coefInt.Sign() > 0
}

// ============================================================
// Helper Functions
// ============================================================

// intToCoef converts a big.Int to 16-byte two's complement representation.
func intToCoef(value *big.Int) [16]byte {
	var result [16]byte
	
	if value.Sign() >= 0 {
		// Positive or zero: direct conversion
		bytes := value.Bytes()
		if len(bytes) > 16 {
			// Truncate (loss of precision)
			bytes = bytes[len(bytes)-16:]
		}
		copy(result[16-len(bytes):], bytes)
	} else {
		// Negative: two's complement
		// value + 2^128 gives the two's complement representation
		temp := new(big.Int).Add(value, new(big.Int).Lsh(big.NewInt(1), 128))
		bytes := temp.Bytes()
		if len(bytes) <= 16 {
			copy(result[16-len(bytes):], bytes)
		}
	}
	
	return result
}

// coefToInt converts a 16-byte two's complement representation to big.Int.
func coefToInt(coef []byte) *big.Int {
	if len(coef) != 16 {
		return big.NewInt(0)
	}
	
	// Check sign bit (MSB of first byte)
	negative := coef[0]&0x80 != 0
	
	result := new(big.Int).SetBytes(coef)
	
	if negative {
		// Two's complement: subtract 2^128
		result.Sub(result, new(big.Int).Lsh(big.NewInt(1), 128))
	}
	
	return result
}

// DecimalFromAny creates a Decimal128 from various types.
func DecimalFromAny(value interface{}) (Decimal128, error) {
	switch v := value.(type) {
	case Decimal128:
		return v, nil
	case string:
		return NewDecimal128FromString(v)
	case int:
		return NewDecimal128FromInt64(int64(v)), nil
	case int64:
		return NewDecimal128FromInt64(v), nil
	case float64:
		return NewDecimal128FromFloat64(v)
	default:
		return Decimal128{}, fmt.Errorf("cannot convert %T to Decimal128", value)
	}
}

// ParseDecimalRegex matches decimal literals with "m" suffix (e.g., "99.99m")
var ParseDecimalRegex = regexp.MustCompile(`^(-?\d+(?:\.\d+)?)m$`)

// IsDecimalLiteral checks if a string is a decimal literal with "m" suffix.
func IsDecimalLiteral(s string) bool {
	return ParseDecimalRegex.MatchString(s)
}

// ParseDecimalLiteral parses a decimal literal with "m" suffix.
func ParseDecimalLiteral(s string) (Decimal128, error) {
	matches := ParseDecimalRegex.FindStringSubmatch(s)
	if len(matches) != 2 {
		return Decimal128{}, fmt.Errorf("not a valid decimal literal: %s", s)
	}
	return NewDecimal128FromString(matches[1])
}

File: attic/rust/glyph-codec/src/decimal128.rs
//! Decimal128 - High-precision decimal type for GLYPH
//!
//! A 128-bit decimal for financial, scientific, and precise mathematical calculations.
//! Value = coefficient * 10^(-scale) where scale is -127 to 127.
//!
//! Unlike f64, Decimal128:
//! - Preserves exact decimal representation
//! - No precision loss for large numbers (>2^53)
//! - Safe for financial calculations
//! - Compatible with blockchain/crypto systems

use std::cmp::Ordering;
use std::fmt;
use std::ops::{Add, Sub, Mul, Div, Neg};
use std::str::FromStr;

const MAX_I128_POW10_EXP: u32 = 38;

/// Decimal128 represents a 128-bit decimal number.
/// Value = coefficient * 10^(-scale)
#[derive(Clone, Copy, Eq)]
pub struct Decimal128 {
    /// Exponent: -127 to 127
    pub scale: i8,
    /// 128-bit coefficient (two's complement, big-endian)
    pub coef: [u8; 16],
}

impl Decimal128 {
    /// Create a new Decimal128 from scale and coefficient bytes.
    pub fn new(scale: i8, coef: [u8; 16]) -> Self {
        Self { scale, coef }
    }

    /// Create a Decimal128 from an i64.
    pub fn from_i64(value: i64) -> Self {
        let mut coef = [0u8; 16];
        let bytes = value.to_be_bytes();

        // Sign extend for negative numbers
        let fill = if value < 0 { 0xFF } else { 0x00 };
        for i in 0..8 {
            coef[i] = fill;
        }
        coef[8..16].copy_from_slice(&bytes);

        Self { scale: 0, coef }
    }

    /// Create a Decimal128 from a string representation.
    pub fn from_string(s: &str) -> Result<Self, DecimalError> {
        let s = s.trim();

        // Remove 'm' suffix if present
        let s = s.strip_suffix('m').unwrap_or(s);

        // Check for negative
        let (negative, s) = if s.starts_with('-') {
            (true, &s[1..])
        } else {
            (false, s)
        };

        // Split by decimal point
        let parts: Vec<&str> = s.split('.').collect();
        if parts.len() > 2 {
            return Err(DecimalError::InvalidFormat);
        }

        let (int_part, frac_part, scale_len) = if parts.len() == 2 {
            (parts[0], parts[1], parts[1].len())
        } else {
            (parts[0], "", 0usize)
        };

        if scale_len > 127 {
            return Err(DecimalError::ScaleOverflow);
        }
        let scale = scale_len as i8;

        // Parse coefficient
        let coef_str = format!("{}{}", int_part, frac_part);
        let mut coef_value = i128::from_str(&coef_str)
            .map_err(|_| DecimalError::InvalidFormat)?;

        if negative {
            coef_value = -coef_value;
        }

        Ok(Self {
            scale,
            coef: int_to_coef(coef_value),
        })
    }

    /// Create a Decimal128 from an f64 (with potential precision loss).
    pub fn from_f64(f: f64) -> Result<Self, DecimalError> {
        Self::from_string(&format!("{}", f))
    }

    /// Convert to i64 (truncates fractional part).
    pub fn to_i64(&self) -> i64 {
        let coef = coef_to_int(&self.coef);
        let value = if self.scale >= 0 {
            match checked_pow10(self.scale as u32) {
                Some(divisor) => coef / divisor,
                None => 0,
            }
        } else {
            match checked_scale_coef(coef, self.scale.unsigned_abs() as u32) {
                Ok(value) => value,
                Err(_) => {
                    return if coef.is_negative() {
                        i64::MIN
                    } else {
                        i64::MAX
                    };
                }
            }
        };

        saturating_i128_to_i64(value)
    }

    /// Convert to f64 (with potential precision loss).
    pub fn to_f64(&self) -> f64 {
        let coef = coef_to_int(&self.coef);
        let divisor = 10f64.powi(self.scale as i32);
        (coef as f64) / divisor
    }

    /// Check if value is zero.
    pub fn is_zero(&self) -> bool {
        coef_to_int(&self.coef) == 0
    }

    /// Check if value is negative.
    pub fn is_negative(&self) -> bool {
        coef_to_int(&self.coef) < 0
    }

    /// Check if value is positive.
    pub fn is_positive(&self) -> bool {
        coef_to_int(&self.coef) > 0
    }

    /// Return the absolute value.
    pub fn abs(&self) -> Self {
        let coef = coef_to_int(&self.coef);
        Self {
            scale: self.scale,
            coef: int_to_coef(coef.wrapping_abs()),
        }
    }

    /// Negate the value.
    pub fn negate(&self) -> Self {
        let coef = coef_to_int(&self.coef);
        Self {
            scale: self.scale,
            coef: int_to_coef(coef.wrapping_neg()),
        }
    }
}

impl fmt::Display for Decimal128 {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let coef = coef_to_int(&self.coef);
        let mut digits = coef.to_string();
        let negative = digits.starts_with('-');
        if negative {
            digits.remove(0);
        }

        let rendered = if self.scale > 0 {
            let mut digits = digits;
            while digits.len() <= self.scale as usize {
                digits.insert(0, '0');
            }
            let insert_pos = digits.len() - self.scale as usize;
            digits.insert(insert_pos, '.');
            digits
        } else if self.scale < 0 {
            let zeros = "0".repeat(self.scale.unsigned_abs() as usize);
            format!("{}{}", digits, zeros)
        } else {
            digits
        };

        if negative {
            write!(f, "-{}", rendered)
        } else {
            write!(f, "{}", rendered)
        }
    }
}

impl fmt::Debug for Decimal128 {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Decimal128(scale={}, value={})", self.scale, self)
    }
}

impl PartialEq for Decimal128 {
    fn eq(&self, other: &Self) -> bool {
        self.cmp(other) == Ordering::Equal
    }
}

impl PartialOrd for Decimal128 {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Decimal128 {
    fn cmp(&self, other: &Self) -> Ordering {
        let c1 = coef_to_int(&self.coef);
        let c2 = coef_to_int(&other.coef);

        if c1 == c2 && self.scale == other.scale {
            return Ordering::Equal;
        }

        if c1.is_negative() != c2.is_negative() {
            return c1.cmp(&c2);
        }

        let ordering = compare_abs_decimal(c1, self.scale, c2, other.scale);
        if c1.is_negative() {
            ordering.reverse()
        } else {
            ordering
        }
    }
}

impl Add for Decimal128 {
    type Output = Result<Self, DecimalError>;

    fn add(self, other: Self) -> Self::Output {
        let target_scale = self.scale.max(other.scale);
        let c1 = align_coef_to_scale(coef_to_int(&self.coef), self.scale, target_scale)?;
        let c2 = align_coef_to_scale(coef_to_int(&other.coef), other.scale, target_scale)?;
        let result = c1.checked_add(c2).ok_or(DecimalError::Overflow)?;

        Ok(Self {
            scale: target_scale,
            coef: int_to_coef(result),
        })
    }
}

impl Sub for Decimal128 {
    type Output = Result<Self, DecimalError>;

    fn sub(self, other: Self) -> Self::Output {
        let target_scale = self.scale.max(other.scale);
        let c1 = align_coef_to_scale(coef_to_int(&self.coef), self.scale, target_scale)?;
        let c2 = align_coef_to_scale(coef_to_int(&other.coef), other.scale, target_scale)?;
        let result = c1.checked_sub(c2).ok_or(DecimalError::Overflow)?;

        Ok(Self {
            scale: target_scale,
            coef: int_to_coef(result),
        })
    }
}

impl Mul for Decimal128 {
    type Output = Result<Self, DecimalError>;

    fn mul(self, other: Self) -> Self::Output {
        let c1 = coef_to_int(&self.coef);
        let c2 = coef_to_int(&other.coef);
        let result = c1.checked_mul(c2).ok_or(DecimalError::Overflow)?;

        let new_scale = self.scale as i16 + other.scale as i16;
        if new_scale > 127 || new_scale < -127 {
            return Err(DecimalError::ScaleOverflow);
        }

        Ok(Self {
            scale: new_scale as i8,
            coef: int_to_coef(result),
        })
    }
}

impl Div for Decimal128 {
    type Output = Result<Self, DecimalError>;

    fn div(self, other: Self) -> Self::Output {
        let c2 = coef_to_int(&other.coef);
        if c2 == 0 {
            return Err(DecimalError::DivisionByZero);
        }

        let c1 = coef_to_int(&self.coef);
        let result = c1.checked_div(c2).ok_or(DecimalError::Overflow)?;

        let new_scale = self.scale as i16 - other.scale as i16;
        if new_scale > 127 || new_scale < -127 {
            return Err(DecimalError::ScaleOverflow);
        }

        Ok(Self {
            scale: new_scale as i8,
            coef: int_to_coef(result),
        })
    }
}

impl Neg for Decimal128 {
    type Output = Self;

    fn neg(self) -> Self::Output {
        self.negate()
    }
}

impl FromStr for Decimal128 {
    type Err = DecimalError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Self::from_string(s)
    }
}

// ============================================================
// Error Type
// ============================================================

/// Decimal128 error types.
#[derive(Debug, Clone, PartialEq)]
pub enum DecimalError {
    InvalidFormat,
    Overflow,
    ScaleOverflow,
    DivisionByZero,
}

impl fmt::Display for DecimalError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            DecimalError::InvalidFormat => write!(f, "invalid decimal format"),
            DecimalError::Overflow => write!(f, "decimal overflow"),
            DecimalError::ScaleOverflow => write!(f, "scale overflow"),
            DecimalError::DivisionByZero => write!(f, "division by zero"),
        }
    }
}

impl std::error::Error for DecimalError {}

// ============================================================
// Helper Functions
// ============================================================

/// Convert i128 to 16-byte two's complement representation.
fn int_to_coef(value: i128) -> [u8; 16] {
    value.to_be_bytes()
}

/// Convert 16-byte two's complement to i128.
fn coef_to_int(coef: &[u8; 16]) -> i128 {
    i128::from_be_bytes(*coef)
}

fn checked_pow10(exp: u32) -> Option<i128> {
    if exp > MAX_I128_POW10_EXP {
        return None;
    }

    let mut value = 1i128;
    for _ in 0..exp {
        value = value.checked_mul(10)?;
    }
    Some(value)
}

fn checked_scale_coef(coef: i128, exp: u32) -> Result<i128, DecimalError> {
    if exp == 0 || coef == 0 {
        return Ok(coef);
    }

    let factor = checked_pow10(exp).ok_or(DecimalError::Overflow)?;
    coef.checked_mul(factor).ok_or(DecimalError::Overflow)
}

fn align_coef_to_scale(coef: i128, from_scale: i8, to_scale: i8) -> Result<i128, DecimalError> {
    let diff = i16::from(to_scale) - i16::from(from_scale);
    if diff < 0 {
        return Err(DecimalError::ScaleOverflow);
    }

    checked_scale_coef(coef, diff as u32)
}

fn normalize_decimal_parts(coef: i128, scale: i8) -> (String, i32) {
    if coef == 0 {
        return ("0".to_string(), 0);
    }

    let mut digits = coef.to_string();
    if digits.starts_with('-') {
        digits.remove(0);
    }
    let mut scale = i32::from(scale);

    while digits.len() > 1 && digits.ends_with('0') {
        digits.pop();
        scale -= 1;
    }

    (digits, scale)
}

fn compare_abs_decimal(left_coef: i128, left_scale: i8, right_coef: i128, right_scale: i8) -> Ordering {
    let (mut left_digits, left_scale) = normalize_decimal_parts(left_coef, left_scale);
    let (mut right_digits, right_scale) = normalize_decimal_parts(right_coef, right_scale);

    let target_scale = left_scale.max(right_scale);
    if target_scale > left_scale {
        left_digits.push_str(&"0".repeat((target_scale - left_scale) as usize));
    }
    if target_scale > right_scale {
        right_digits.push_str(&"0".repeat((target_scale - right_scale) as usize));
    }

    left_digits
        .len()
        .cmp(&right_digits.len())
        .then_with(|| left_digits.cmp(&right_digits))
}

fn saturating_i128_to_i64(value: i128) -> i64 {
    if value > i64::MAX as i128 {
        i64::MAX
    } else if value < i64::MIN as i128 {
        i64::MIN
    } else {
        value as i64
    }
}

/// Check if a string is a decimal literal (ends with 'm').
pub fn is_decimal_literal(s: &str) -> bool {
    let s = s.trim();
    if !s.ends_with('m') {
        return false;
    }
    let num_part = &s[..s.len()-1];
    Decimal128::from_string(num_part).is_ok()
}

/// Parse a decimal literal (with 'm' suffix).
pub fn parse_decimal_literal(s: &str) -> Result<Decimal128, DecimalError> {
    let s = s.trim();
    if !s.ends_with('m') {
        return Err(DecimalError::InvalidFormat);
    }
    Decimal128::from_string(&s[..s.len()-1])
}

// ============================================================
// Tests
// ============================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_from_i64() {
        let d = Decimal128::from_i64(123);
        assert_eq!(d.to_string(), "123");
        assert_eq!(d.to_i64(), 123);

        let d = Decimal128::from_i64(-456);
        assert_eq!(d.to_string(), "-456");
        assert_eq!(d.to_i64(), -456);
    }

    #[test]
    fn test_from_string() {
        let d = Decimal128::from_string("123.45").unwrap();
        assert_eq!(d.to_string(), "123.45");
        assert_eq!(d.scale, 2);

        let d = Decimal128::from_string("-99.99").unwrap();
        assert_eq!(d.to_string(), "-99.99");

        let d = Decimal128::from_string("0.0001").unwrap();
        assert_eq!(d.to_string(), "0.0001");
    }

    #[test]
    fn test_negative_scale_display_and_conversion() {
        let d = Decimal128::new(-2, int_to_coef(123));
        assert_eq!(d.to_string(), "12300");
        assert_eq!(d.to_i64(), 12300);

        let neg = Decimal128::new(-3, int_to_coef(-45));
        assert_eq!(neg.to_string(), "-45000");
        assert_eq!(neg.to_i64(), -45000);
    }

    #[test]
    fn test_arithmetic() {
        let d1 = Decimal128::from_string("100.50").unwrap();
        let d2 = Decimal128::from_string("50.25").unwrap();

        let sum = (d1 + d2).unwrap();
        assert_eq!(sum.to_string(), "150.75");

        let diff = (d1 - d2).unwrap();
        assert_eq!(diff.to_string(), "50.25");
    }

    #[test]
    fn test_comparison() {
        let d1 = Decimal128::from_string("100.50").unwrap();
        let d2 = Decimal128::from_string("100.50").unwrap();
        let d3 = Decimal128::from_string("50.25").unwrap();

        assert_eq!(d1, d2);
        assert!(d1 > d3);
        assert!(d3 < d1);
    }

    #[test]
    fn test_predicates() {
        let zero = Decimal128::from_i64(0);
        assert!(zero.is_zero());

        let neg = Decimal128::from_string("-10.5").unwrap();
        assert!(neg.is_negative());

        let pos = Decimal128::from_string("10.5").unwrap();
        assert!(pos.is_positive());
    }

    #[test]
    fn test_decimal_literal() {
        assert!(is_decimal_literal("99.99m"));
        assert!(is_decimal_literal("100m"));
        assert!(!is_decimal_literal("99.99"));

        let d = parse_decimal_literal("99.99m").unwrap();
        assert_eq!(d.to_string(), "99.99");
    }

    #[test]
    fn test_add_overflow_returns_error() {
        let max = Decimal128::new(0, int_to_coef(i128::MAX));
        let one = Decimal128::from_i64(1);

        assert_eq!(max + one, Err(DecimalError::Overflow));
    }

    #[test]
    fn test_mul_overflow_returns_error() {
        let huge = Decimal128::new(0, int_to_coef(i128::MAX));
        let two = Decimal128::from_i64(2);

        assert_eq!(huge * two, Err(DecimalError::Overflow));
    }

    #[test]
    fn test_compare_negative_scale() {
        let lhs = Decimal128::new(-2, int_to_coef(123));
        let rhs = Decimal128::from_i64(12299);

        assert!(lhs > rhs);
    }
}
